File: frontend/src/types/GameInstance.ts
import { GameCard } from './GameCard'
import { GameNode } from './GameNode'

export type GameInstanceStatus = 'running' | 'stopped' | 'error'

export interface GameInstanceConfig {
  maxPlayers: number
  port: number
  settings: {
    map: string
    difficulty: string
  }
}

export interface GameInstanceMetrics {
  cpuUsage: number
  memoryUsage: number
  storageUsage: number
  gpuUsage?: number
  networkUsage: number
  uptime: number
  fps: number
  playerCount: number
}

export interface GameInstance {
  id: string
  gameId: string
  gameCardId: string
  gameCard: {
    id: string
    name: string
    platform: {
      id: string
      name: string
      type: string
      status: string
      description: string
      version: string
      image: string
      config: {
        apiKey: string
        apiUrl: string
        callbackUrl: string
      }
      environment: {
        dockerImage: string
        envVars: Record<string, string>
      }
      features: {
        autoScale: boolean
        backup: boolean
        monitoring: boolean
      }
      createdAt: string
      updatedAt: string
    }
    description: string
    coverImage: string
    status: string
    createdAt: string
  }
  nodeId: string
  node: {
    id: string
    name: string
    description: string
    type: string
    status: string
    region: string
    network: {
      ip: string
      port: number
      protocol: string
      bandwidth: number
    }
    resources: {
      cpu: number
      memory: number
      storage: number
      network: number
      gpu?: {
        model: string
        memory: number
      }
    }
    metrics: {
      cpuUsage: number
      memoryUsage: number
      storageUsage: number
      networkUsage: number
      gpuUsage?: number
      uptime: number
      fps: number
      instanceCount: number
      playerCount: number
    }
    labels: Record<string, string>
    createdAt: string
    updatedAt: string
  }
  name: string
  status: GameInstanceStatus
  config: GameInstanceConfig
  metrics: GameInstanceMetrics
  logs: string[]
  createdAt: string
  updatedAt: string
}

export interface GameInstanceQuery {
  page: number
  pageSize: number
  keyword?: string
  gameCardId?: string
  nodeId?: string
  status?: string
}

export interface GameInstanceForm {
  id?: string
  gameCardId: string
  nodeId: string
  config: {
    maxPlayers: number
    port: number
    settings: Record<string, any>
  }
} 

File: internal/models/gameinstance.go
package models

import "time"

// GameInstance 游戏实例模型
type GameInstance struct {
	ID          string    `json:"id" yaml:"id"`
	NodeID      string    `json:"node_id" yaml:"node_id"`         // 关联的游戏机ID
	PlatformID  string    `json:"platform_id" yaml:"platform_id"` // 关联的平台ID
	CardID      string    `json:"card_id" yaml:"card_id"`         // 关联的游戏卡片ID
	Status      string    `json:"status" yaml:"status"`           // 运行状态
	Resources   string    `json:"resources" yaml:"resources"`     // 资源占用
	Performance string    `json:"performance" yaml:"performance"` // 性能指标
	SaveData    string    `json:"save_data" yaml:"save_data"`     // 存档数据
	Config      string    `json:"config" yaml:"config"`           // 实例配置
	Backup      string    `json:"backup" yaml:"backup"`           // 备份数据
	CreatedAt   time.Time `json:"created_at" yaml:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" yaml:"updated_at"`
	StartedAt   time.Time `json:"started_at" yaml:"started_at"` // 启动时间
	StoppedAt   time.Time `json:"stopped_at" yaml:"stopped_at"` // 停止时间
}

// TableName 返回表名
func (GameInstance) TableName() string {
	return "game_instances"
}
